### PR TITLE
[3.12] gh-113993: Make interned strings mortal (GH-120520, GH-121364, GH-121903, GH-122303)

### PR DESCRIPTION
**Pull request by encukou**
This backports several PRs for gh-113993, making interned strings mortal so they can be garbage-collected when no longer needed.

* Allow interned strings to be mortal, and fix related issues (GH-120520)

  * Add an InternalDocs file describing how interning should work and how to use it.

  * Add internal functions to *explicitly* request what kind of interning is done:
    - `_PyUnicode_InternMortal`
    - `_PyUnicode_InternImmortal`
    - `_PyUnicode_InternStatic`

  * Switch uses of `PyUnicode_InternInPlace` to those.

  * Disallow using `_Py_SetImmortal` on strings directly.
    You should use `_PyUnicode_InternImmortal` instead:
    - Strings should be interned before immortalization, otherwise you're possibly
      interning a immortalizing copy.
    - `_Py_SetImmortal` doesn't handle the `SSTATE_INTERNED_MORTAL` to
      `SSTATE_INTERNED_IMMORTAL` update, and those flags can't be changed in
      backports, as they are now part of public API and version-specific ABI.

  * Add private `_only_immortal` argument for `sys.getunicodeinternedsize`, used in refleak test machinery.

   Make sure the statically allocated string singletons are unique. This means these sets are now disjoint:
    - `_Py_ID`
    - `_Py_STR` (including the empty string)
    - one-character latin-1 singletons

    Now, when you intern a singleton, that exact singleton will be interned.

  * Add a `_Py_LATIN1_CHR` macro, use it instead of `_Py_ID`/`_Py_STR` for one-character latin-1 singletons everywhere (including Clinic).

  * Intern `_Py_STR` singletons at startup.

  * Beef up the tests. Cover internal details (marked with `@cpython_only`).

  * Add lots of assertions

* Don't immortalize in PyUnicode_InternInPlace; keep immortalizing in other API (GH-121364)

  * Switch PyUnicode_InternInPlace to _PyUnicode_InternMortal, clarify docs

  * Document immortality in some functions that take `const char *`

  This is PyUnicode_InternFromString;
  PyDict_SetItemString, PyObject_SetAttrString;
  PyObject_DelAttrString; PyUnicode_InternFromString;
  and the PyModule_Add convenience functions.

  Always point out a non-immortalizing alternative.

  * Don't immortalize user-provided attr names in _ctypes

* Immortalize names in code objects to avoid crash (GH-121903)

* Intern latin-1 one-byte strings at startup (GH-122303)

There are some 3.12-specific changes, mainly to allow statically allocated strings in deepfreeze. (In 3.13, deepfreeze switched to the general `_Py_ID`/`_Py_STR`.)

Co-authored-by: Eric Snow <ericsnowcurrently@gmail.com>

----
📚 Documentation preview 📚: https://cpython-previews--123065.org.readthedocs.build/

Issue: #113993

### CODE REVIEW

**Comment by encukou** [comment]
The buildbot failures look unrelated.

**Comment by encukou** [comment]
@Yhg1s, please review this backport. Sorry about the size!

File: InternalDocs/string_interning.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+- the deallocator (`unicode_dealloc`) removes the string from the interned dict
+- at shutdown, when the interned dict is cleared, the references are added back
+
+As with any type, you should only immortalize strings that will live until
+interpreter shutdown.
+We currently also immortalize strings contained in code objects and similar,
+specifically in the compiler and in `marshal`.
+These are “close enough” to immortal: even in use cases like hot reloading
+or `eval`-ing user input, the number of distinct identifiers and string
+constants expected to stay low.
```
Comment (JelleZijlstra) [inline InternalDocs/string_interning.md:79]:
```suggestion
constants is expected to stay low.
```

File: Tools/build/generate_global_objects.py [hunk 2]
```diff
@@ -406,15 +411,31 @@ def generate_global_object_finalizers(generated_immortal_objects):
...
     for name, string, *_ in iter_global_strings():
         if string is None:
             if name not in IGNORED:
                 identifiers.add(name)
         else:
+            if len(string) == 1 and ord(string) < 256:
+                # Give a nice message for common mistakes.
+                # To cover tricky cases (like "\n") we also generate C asserts.
+                raise ValueError(
+                    'do not use &_PyID or &_Py_STR for one-character latin-1 '
```
Comment (JelleZijlstra) [inline Tools/build/generate_global_objects.py]:
```suggestion
                    'do not use &_Py_ID or &_Py_STR for one-character latin-1 '
```

File: Tools/build/generate_global_objects.py [hunk 3]
```diff
@@ -406,15 +411,31 @@ def generate_global_object_finalizers(generated_immortal_objects):
...
+                    + f'strings, use _Py_LATIN1_CHR instead: {string!r}')
             if string not in strings:
                 strings[string] = name
             elif name != strings[string]:
                 raise ValueError(f'string mismatch for {name!r} ({string!r} != {strings[name]!r}')
+    overlap = identifiers & set(strings.keys())
+    if overlap:
+        raise ValueError(
+            'do not use both _PyID and _Py_DECLARE_STR for the same string: '
```
Comment (JelleZijlstra) [inline Tools/build/generate_global_objects.py]:
```suggestion
            'do not use both _Py_ID and _Py_DECLARE_STR for the same string: '
```

File: InternalDocs/string_interning.md (hunk 1 above)
Comment (encukou) [reply InternalDocs/string_interning.md:79]: Thanks! I'll fix this main. I don't think it's worth backporting though.
Edit: it's #124699

**Comment by mgorny** [comment]
FYI, this is at least missing a backport of 281ffb60cc194c4eefee3e112888251abd1a749a, and therefore causing Rust packages to crash on assertions. I've filed #124887.

**Pull request merged**